src/compiler/crystal/semantic/flags.cr

```diff
@@ -27,6 +27,7 @@ class Crystal::Program
     set.add "freebsd" if set.any?(&.starts_with?("freebsd"))
     set.add "openbsd" if set.any?(&.starts_with?("openbsd"))
     set.add "unix" if set.any? { |flag| %w(cygnus darwin freebsd linux openbsd).includes?(flag) }
+    set.add "win32" if set.any?(&.starts_with?("windows")) && set.any? { |flag| %w(gnu msvc).includes?(flag) }
```

Comment:
The reason is given in https://github.com/crystal-lang/crystal/pull/4832#issuecomment-323078139